src/index.ts

```diff
@@ -59,7 +59,7 @@ export default function createClient<T>(defaultOptions?: ClientOptions) {
     : { params?: BaseParams };
   type Params<T> = PathParams<T> & MethodParams<T>;
   type RequestBody<T> = T extends { requestBody: any } ? { body: Unwrap<T['requestBody']> } : { body?: never };
-  type FetchOptions<T> = Params<T> & RequestBody<T> & Omit<RequestInit, 'body'>;
+  type FetchOptions<T> = Params<T> & RequestBody<T> & { querySerializer?: (q: any) => string } & Omit<RequestInit, 'body'>;
```

Comment:
```suggestion
  type FetchOptions<T> = Params<T> & RequestBody<T> & { querySerializer?: (q: Params<T>) => string } & Omit<RequestInit, 'body'>;
```

I think this could also use the available type inferences, so we alert users of any errors?

Reply:
I’m personally not a fan of the proposed API of requiring the user to assemble a string themselves.

What about this API instead?

```ts
{ querySerializer?: (q: Params<T>) => Record<string, string> }
```

e.g.

```diff
  await client.get('/post/{post_id}', {
    params: { path: { post_id: 'my-post' }, query: { a: 1, b: 2 } },
-   querySerializer: (q) => `alpha=${q.a}&beta=${q.b}`,
+   querySerializer: (a) => ({ alpha: q.a, beta: q.b })
  });
```

We then take that object and pass it back into `URLSearchParams()` so that handles string mashing and URL encoding. IMO requiring a user to take on that task can get annoying when the goal is only renaming/remapping of parameters

Reply:
I‘d actually be totally fine with stitching together that string. I‘m just not sure that doing it per request makes sense. I may be completely mistaken here, but couldn’t this be passed to `createClient` once? There may be cases where you‘d want to adjust the serializer per request. But my guess is most APIs (hopefully…) use one style of serializing queries across all routes 😅

Reply:
> I’d actually be totally fine with stitching together that string.

That’s good input, thank you! As I thought about it more, I’m more and more in favor of the original PR implementation and take back what I said. I think in many scenarios, devs probably wouldn’t be using string templating, and instead they’d provide their own serializer methods [based on their preferred style](https://swagger.io/docs/specification/serialization/)

> couldn’t this be passed to `createClient` once? … my guess is most APIs (hopefully…) use one style of serializing queries

I’ve worked on `openapi-typescript` long enough to know people do some _silly_ things with their API specs 🤪. Since param serialization isn’t necessary dictated by the spec, I think it would be wise to support different serialization styles at the request level (e.g. stitched APIs).

But that said, I think we could also allow it on the top `createClient` level as well. And users could override that per-request if needed. Runtime-wise, it will still get called on every request so it’d be the same performance cost either way.